### PR TITLE
Consensus

### PR DESCRIPTION
**Pull request by kousthubraja**

### CODE REVIEW

File: app/codes/p2p/sync_chain.py [hunk 1]
```diff
@@ -29,10 +34,24 @@ def receive_block(block):
...
     return True
 
 
+def receive_receipt(receipt):
+    logger.info('Recieved receipt: %s', receipt)
+    if not validate_receipt_signature(receipt):
+        logger.info('Invalid receipt signature')
+        return False
+
+    # TODO - Add the receipt to an existing block in Temp folder
```
Comment (kousthubraja) [inline app/codes/p2p/sync_chain.py:48]: @swapnilpawar31415 is this logic for temp receipt storage right?

File: app/codes/validator.py [hunk 2]
```diff
@@ -45,3 +50,70 @@ def validate(transaction):
...
+        total_receipt_count += 1
+
+        if not validate_receipt_signature(receipt):
+            continue
+
+        if receipt['data']['block_index'] != block['index'] or receipt['data']['block_hash'] != block['hash'] or receipt['data']['vote'] < 1:
+            continue
+
+        trust_score = get_node_trust_score(receipt['public_key'])
+        valid_probability = 0 if trust_score < 0 else (trust_score + 2) / 5
```
Comment (kousthubraja) [inline app/codes/validator.py:93]: @swapnilpawar31415 critical part here

File: app/codes/p2p/sync_chain.py [hunk 3]
```diff
@@ -29,10 +34,24 @@ def receive_block(block):
     block_index = block['block_index'] if 'block_index' in block else block['index']
     if block_index > get_last_block_index() + 1:
         sync_chain_from_peers()
+    validate_block(block)
```
Comment (swapnilpawar31415) [inline app/codes/p2p/sync_chain.py:37]: Should we add a condition here?
flag = validate_block(block)
if not flag:
    print("Block is not valid")  #can also be logger
    return False

File: app/codes/p2p/sync_chain.py (hunk 1 above)
Comment (swapnilpawar31415) [reply app/codes/p2p/sync_chain.py:48]: Yes.

File: app/codes/validator.py (hunk 2 above)
Comment (swapnilpawar31415) [reply app/codes/validator.py:93]: Will re-write the logic in the white-paper for this part and send to you.
Need to separate the information contained in a negative vote vs no vote, which is not achieved currently.

File: app/codes/validator.py [hunk 4]
```diff
@@ -45,3 +50,70 @@ def validate(transaction):
...
+
+        score_weighted_validity_count += valid_probability
+
+    if score_weighted_validity_count < 0.75:
```
Comment (swapnilpawar31415) [inline app/codes/validator.py:97]: add a constant ACCEPTANCE_THRESHOLD and use it instead of 0.75

**Pull request closed without merging**